packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2806,12 +2811,23 @@ export class ContainerRuntime
 				runtimeBatch,
 			);
 		} else {
+			if (!runtimeBatch) {
+				// The DeltaManager used to do this, but doesn't anymore as of Loader v2.4
+				// Anyone listening to our "op" event would expect the contents to be parsed per this same logic
+				if (
+					typeof messageCopy.contents === "string" &&
+					messageCopy.contents !== "" &&
+					messageCopy.type !== MessageType.ClientLeave
+				) {
+					messageCopy.contents = JSON.parse(messageCopy.contents);
```

Comment:
Why not use `ensureContentsDeserialized` above for this case as well? Is it just because of the `messageCopy.type !== MessageType.ClientLeave` condition? That messages is handled in the `Loader` layer so it doesn't matter.

Reply:
> Why not use ensureContentsDeserialized above for this case as well? 

It's to keep the two codepaths separate, since the scenario is different (actually processing runtime messages which this code has "ownership" of, v. observing non-runtime messages and emitting the "op" event). We must deserialize a runtime message's contents, but we could choose to emit "op" event for non-runtime messages with contents as-is (or even stop emitting that event altogether).  And, yes, we can skip the `MessageType.ClientLeave` check in `ensureContentsDeserialized`.

> That messages is handled in the Loader layer so it doesn't matter

But those messages come here too.  Maybe it's a non-JSON string and would crash.